grafast/grafserv/src/servers/lambda/index.ts

```diff
@@ -0,0 +1,132 @@
+import type {
+  APIGatewayEvent,
+  APIGatewayProxyEvent,
+  APIGatewayProxyResult,
+  Context as LambdaContext,
+} from "aws-lambda";
+import type {
+  GrafservConfig,
+  RequestDigest} from "grafserv";
+import {
+  GrafservBase,
+  processHeaders
+} from "grafserv";
+import { toArray } from "modern-async";
+
+export class LambdaGrafserv extends GrafservBase {
+  createHandler() {
+    // eslint-disable-next-line @typescript-eslint/no-this-alias
+    const serv = this;
+    return async function (
+      event: APIGatewayEvent,
+      context: LambdaContext,
+    ): Promise<APIGatewayProxyResult> {
+      const version = event.requestContext.protocol.match(
+        "^HTTP/(?<major>[0-9]+).(?<minor>[0-9]+)$",
+      );
+
+      const request: RequestDigest = {
+        httpVersionMajor: parseInt(version?.groups?.major ?? "1"),
+        httpVersionMinor: parseInt(version?.groups?.minor ?? "0"),
+        isSecure: event.headers["X-Forwarded-Proto"] === "https",
+        method: event.httpMethod,
+        path: event.requestContext.path,
+        headers: processHeaders(event.multiValueHeaders),
+        getQueryParams() {
+          return Object.fromEntries(
+            Object.entries(event.queryStringParameters ?? {}).filter(
+              ([_k, v]) => v !== undefined,
+            ),
+          ) as Record<string, string>;
+        },
+        getBody() {
+          return {
+            type: "text",
+            text: event.body ?? "",
+          };
+        },
+        requestContext: {
+          lambdaCtx: { event, context },
```

Comment:
I was mislead 😅

```
     * It's anticipated this will be expanded via declaration merging, e.g. if
     * your server is Koa then a `koaCtx` might be added.
```

Reply:
https://github.com/graphile/crystal/pull/1783